compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -396,3 +400,24 @@ impl<'a> ast::mut_visit::MutVisitor for TypeSubstitution<'a> {
         }
     }
 }
+
+struct DetectNonGenericPointeeAttr<'a, 'b> {
+    cx: &'a ExtCtxt<'b>,
+}
+
+impl<'a, 'b> rustc_ast::visit::Visitor<'a> for DetectNonGenericPointeeAttr<'a, 'b> {
+    fn visit_attribute(&mut self, attr: &'a rustc_ast::Attribute) -> Self::Result {
+        if attr.has_name(sym::pointee) {
+            self.cx.dcx().emit_err(errors::NonGenericPointee { span: attr.span });
+        }
+    }
+
+    fn visit_generic_param(&mut self, param: &'a rustc_ast::GenericParam) -> Self::Result {
+        match param.kind {
+            GenericParamKind::Type { .. } => return,
```

Comment:
Hmm.

Technically I think you still need to ensure you recur into the `default: &Option<P<Ty>>` that's held within that enum payload above.

I think the below is the kind of case you need to worry about that motivates recurring here:

```rust
struct SomeStruct<'a, T = [u32; const { #[pointee] struct UhOhYetAnotherTypeDefn<U>; 10 }]> { ... }
```

(Yes its a corner case.)

Reply:
You can model the recursion you need to do by looking at `walk_generic_param`, and just leaving out the `visit_attribute` part:

https://github.com/rust-lang/rust/blob/c39f318c5e734e7f254c6b2589205d6ed936443d/compiler/rustc_ast/src/visit.rs#L749-L767

Reply:
I've fixed the edge case. As a comment in the new code mentions we end up rejecting valid code like:
```rust
#![feature(derive_smart_pointer)]
use std::marker::SmartPointer;
#[derive(SmartPointer)]
#[repr(transparent)]
struct Weird<
    'a,
    T: ?Sized,
    const V: u32 = {
        #[derive(SmartPointer)]
        #[repr(transparent)]
        struct UhOh<'b, #[pointee] X: ?Sized>(&'b X);
        10
    },
> {
    ptr: &'a T,
}
```
Handling this properly isn't possible in the generic case since we can't know if the inner derive is actually a different trait with the same name. Given how unlikely the whole scenario is I've decided to go ahead with the changes anyway.